### DISCUSSION TITLE
auth callback: access_token as query param?

### DISCUSSION THREAD

**Discussion by Schniz** [supabase/supabase#2133] [Questions] [answered]
Hey there!

I wonder -- is there a way to ask the auth backend to redirect back with the access_token in a query param instead of a hash?

i.e.

```diff char
- https://example.com/auth/callback#auth_token=12345
+ https://example.com/auth/callback?auth_token=12345
```

This way, we can use the auth token in a backend, storing it in a session store or whatever, and not only use it in a frontend code. I'm currently playing with simply redirecting to the same URL with the query parameter if I have an `#auth_token` but I don't like this implementation. Feels clunky, and the user gets an unnecessary redirect! -- unless there's a security concern I am not aware of (which in this case, I'd be happy to learn!)

I wasn't sure if I needed to open this as an issue in https://github.com/supabase/gotrue or as a discussion, so I'm genuinely sorry if this is the wrong place!

Thanks for a great product!

**Comment by inian** [comment, marked as answer]
Hi @Schniz, 
It is part of the URL fragment as mandated by the OAuth Spec and we don't plan on changing it for that reason. 
This has been discussed before  here which should give more context - https://github.com/supabase/supabase/issues/1698#issuecomment-847646400

**Reply by Schniz** [reply]
TIL! Thanks!

**Reply by ky1ejs** [reply]
@inian would you be able to offer any clarification here about an implementation in `supabase-community/gotrue-swift` that contradicts this implementation point? 🙏🏼

https://github.com/supabase-community/gotrue-swift/issues/22

**Reply by vibegui** [reply]
Hi @inian 

Sorry to re-beat a dead horse here, but I am building deno, server-side-first applications using supabase (and loving it), but having to kick back to the client in order to "leak the token to myself" and finally `set-cookie` is **extremely** awkward.

It seems like the good people at Auth0 offer the alternative of ["switching to the authorization code grant which would return the code in the query component that would indeed be sent to the server-side"](https://community.auth0.com/t/the-redirect-response-does-not-send-the-hash-fragment-tokens-to-the-server-side/6180).

![CleanShot 2022-11-24 at 11 52 57](https://user-images.githubusercontent.com/1633518/203813099-84bf1c0e-dcb0-4c53-b9c5-a19472013e46.png)

Isn't there a possibility for us to _add that as an option_ so that server-side apps can avoid doing a loop over the network and therefore offer a slower experience?

cc @ky1ejs @Schniz

**Reply by hf** [reply]
Hey @guifromrio,

We are not planning on using the authorization code grant where the code arrives via a query param. We are however going to be working on implementing PKCE, which is the proper replacement for the implicit grant flow.

Should you need access to the access and refresh tokens, use the implicit grant flow to set cookies on the domain(s) you need access to them, and then read those cookies from the server side.
